File: backend/converter.py
import pandas as pd
import json
import os
import re

def formatar_valor(valor, tipo='string'):
    if pd.isna(valor):
        return ""
    
    valor_str = str(valor).strip()
    
    if tipo == 'inteiro':
        return re.sub(r'\.0$', '', valor_str)
        
    return valor_str

diretorio_atual = os.path.dirname(os.path.abspath(__file__))

nome_arquivo_dados = "Relação imobilizados Rio Verde - 27_03_2025 (1).XLSX - Sheet1.csv"
caminho_arquivo_dados = os.path.join(diretorio_atual, nome_arquivo_dados)

nome_arquivo_saida = "dados_imobilizados_processados.json"
caminho_arquivo_saida = os.path.join(diretorio_atual, nome_arquivo_saida)

if not os.path.exists(caminho_arquivo_dados):
    print(f"❌ ERRO: Arquivo de dados não encontrado!")
    print(f"Por favor, verifique se o arquivo '{nome_arquivo_dados}' está na mesma pasta que este script.")
    input("\nPressione Enter para sair...")
    exit()

print(f"ℹ️  Arquivo de dados encontrado: {nome_arquivo_dados}")

try:
    df = pd.read_csv(caminho_arquivo_dados, encoding='utf-8', sep=',')

    colunas_obrigatorias = [
        'Imobilizado', 'Subnº', 'Incorporação em', 
        'Denominação do imobilizado', 'Nº inventário', 
        'Nº de série', 'Centro custo'
    ]
    
    colunas_faltantes = [col for col in colunas_obrigatorias if col not in df.columns]
    if colunas_faltantes:
        print(f"❌ ERRO: Colunas obrigatórias não encontradas no arquivo: {', '.join(colunas_faltantes)}")
        input("\nPressione Enter para sair...")
        exit()

    dados_finais = {}
    linhas_ignoradas = 0
    codigos_invalidos = set()

    for _, row in df.iterrows():
        codigo_imobilizado = formatar_valor(row['Imobilizado'], 'inteiro')
        
        if not codigo_imobilizado or pd.isna(row['Denominação do imobilizado']) or pd.isna(row['Centro custo']):
            if not codigo_imobilizado:
                codigos_invalidos.add(str(row['Imobilizado']))
            linhas_ignoradas += 1
            continue

        try:
            data_formatada = pd.to_datetime(row['Incorporação em']).strftime('%d/%m/%Y')
        except (ValueError, TypeError):
            data_formatada = "Data inválida"

        item = {
            "Subnº": formatar_valor(row['Subnº'], 'inteiro'),
            "Data": data_formatada,
            "Descrição": formatar_valor(row['Denominação do imobilizado']),
            "Inventário": formatar_valor(row['Nº inventário'], 'inteiro'),
            "Série": formatar_valor(row['Nº de série']),
            "Centro Custo": formatar_valor(row['Centro custo'], 'inteiro'),
            "Link": f"https://videplast.com.br/consulta-imobilizado?id={codigo_imobilizado}",
            "QRCODE": ""
        }

        if codigo_imobilizado not in dados_finais:
            dados_finais[codigo_imobilizado] = {"itens": []}
        
        dados_finais[codigo_imobilizado]["itens"].append(item)

    with open(caminho_arquivo_saida, 'w', encoding='utf-8') as f:
        json.dump(dados_finais, f, ensure_ascii=False, indent=4)

    total_codigos = len(dados_finais)
    total_itens = sum(len(v["itens"]) for v in dados_finais.values())
    
    print("\n" + "="*40)
    print("✅ Processamento Concluído com Sucesso!")
    print("="*40)
    print(f"👉 Total de códigos de imobilizado únicos: {total_codigos}")
    print(f"👉 Total de itens (sub-ativos) processados: {total_itens}")
    print(f"👉 Total de linhas ignoradas (dados inválidos): {linhas_ignoradas}")
    
    if codigos_invalidos:
        print(f"⚠️  Amostra de códigos inválidos ignorados: {', '.join(list(codigos_invalidos)[:5])}...")

    print("-"*40)
    print(f"\n🚀 Arquivo JSON gerado com sucesso em:")
    print(caminho_arquivo_saida)
    print("="*40)

except Exception as e:
    print("\n❌ Ocorreu um erro inesperado durante o processamento.")
    print(f"   Tipo do erro: {type(e).__name__}")
    print(f"   Detalhes: {e}")

finally:
    input("\nPressione Enter para sair...")

File: backend/converter.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');
const readline = require('readline');

/**
 * Formata um valor removendo espaços e tratando valores nulos/indefinidos
 * @param {any} valor - Valor a ser formatado
 * @param {string} tipo - Tipo de formatação ('string' ou 'inteiro')
 * @returns {string} Valor formatado
 */
function formatarValor(valor, tipo = 'string') {
    if (valor === null || valor === undefined || valor === '') {
        return "";
    }
    
    const valorStr = String(valor).trim();
    
    if (tipo === 'inteiro') {
        // Remove .0 do final se existir
        return valorStr.replace(/\.0$/, '');
    }
    
    return valorStr;
}

/**
 * Formata uma data para o padrão DD/MM/YYYY
 * @param {string} dataStr - String da data
 * @returns {string} Data formatada ou "Data inválida"
 */
function formatarData(dataStr) {
    if (!dataStr || dataStr.trim() === '') {
        return "Data inválida";
    }
    
    try {
        const data = new Date(dataStr);
        if (isNaN(data.getTime())) {
            return "Data inválida";
        }
        
        const dia = String(data.getDate()).padStart(2, '0');
        const mes = String(data.getMonth() + 1).padStart(2, '0');
        const ano = data.getFullYear();
        
        return `${dia}/${mes}/${ano}`;
    } catch (error) {
        return "Data inválida";
    }
}

/**
 * Aguarda entrada do usuário
 * @param {string} pergunta - Pergunta a ser exibida
 * @returns {Promise<string>} Resposta do usuário
 */
function aguardarEntrada(pergunta) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    return new Promise((resolve) => {
        rl.question(pergunta, (resposta) => {
            rl.close();
            resolve(resposta);
        });
    });
}

/**
 * Função principal de processamento
 */
async function processarDados() {
    const diretorioAtual = __dirname;
    
    const nomeArquivoDados = "imobilizados.csv";
    const caminhoArquivoDados = path.join(diretorioAtual, nomeArquivoDados);
    
    const nomeArquivoSaida = "dados_imobilizados_processados.json";
    const caminhoArquivoSaida = path.join(diretorioAtual, nomeArquivoSaida);
    
    // Verifica se o arquivo de dados existe
    if (!fs.existsSync(caminhoArquivoDados)) {
        console.log("❌ ERRO: Arquivo de dados não encontrado!");
        console.log(`Por favor, verifique se o arquivo '${nomeArquivoDados}' está na mesma pasta que este script.`);
        await aguardarEntrada("\nPressione Enter para sair...");
        process.exit(1);
    }
    
    console.log(`ℹ️  Arquivo de dados encontrado: ${nomeArquivoDados}`);
    
    try {
        const colunasObrigatorias = [
            'Imobilizado', 'Subnº', 'Incorporação em', 
            'Denominação do imobilizado', 'Nº inventário', 
            'Nº de série', 'Centro custo'
        ];
        
        const dadosFinais = {};
        let linhasIgnoradas = 0;
        const codigosInvalidos = new Set();
        let primeiraLinha = true;
        let colunasFaltantes = [];
        
        // Processa o arquivo CSV
        await new Promise((resolve, reject) => {
            fs.createReadStream(caminhoArquivoDados)
                .pipe(csv())
                .on('headers', (headers) => {
                    // Verifica se todas as colunas obrigatórias estão presentes
                    colunasFaltantes = colunasObrigatorias.filter(col => !headers.includes(col));
                    if (colunasFaltantes.length > 0) {
                        reject(new Error(`Colunas obrigatórias não encontradas no arquivo: ${colunasFaltantes.join(', ')}`));
                        return;
                    }
                })
                .on('data', (row) => {
                    const codigoImobilizado = formatarValor(row['Imobilizado'], 'inteiro');
                    
                    // Verifica se os dados obrigatórios estão presentes
                    if (!codigoImobilizado || 
                        !row['Denominação do imobilizado'] || 
                        !row['Centro custo']) {
                        if (!codigoImobilizado) {
                            codigosInvalidos.add(String(row['Imobilizado']));
                        }
                        linhasIgnoradas++;
                        return;
                    }
                    
                    const dataFormatada = formatarData(row['Incorporação em']);
                    
                    const item = {
                        "Subnº": formatarValor(row['Subnº'], 'inteiro'),
                        "Data": dataFormatada,
                        "Descrição": formatarValor(row['Denominação do imobilizado']),
                        "Inventário": formatarValor(row['Nº inventário'], 'inteiro'),
                        "Série": formatarValor(row['Nº de série']),
                        "Centro Custo": formatarValor(row['Centro custo'], 'inteiro'),
                        "Link": `https://videplast.com.br/consulta-imobilizado?id=${codigoImobilizado}`,
                        "QRCODE": ""
                    };
                    
                    if (!dadosFinais[codigoImobilizado]) {
                        dadosFinais[codigoImobilizado] = { "itens": [] };
                    }
                    
                    dadosFinais[codigoImobilizado]["itens"].push(item);
                })
                .on('end', () => {
                    resolve();
                })
                .on('error', (error) => {
                    reject(error);
                });
        });
        
        // Salva o arquivo JSON
        fs.writeFileSync(caminhoArquivoSaida, JSON.stringify(dadosFinais, null, 4), 'utf8');
        
        const totalCodigos = Object.keys(dadosFinais).length;
        const totalItens = Object.values(dadosFinais).reduce((total, obj) => total + obj.itens.length, 0);
        
        console.log("\n" + "=".repeat(40));
        console.log("✅ Processamento Concluído com Sucesso!");
        console.log("=".repeat(40));
        console.log(`👉 Total de códigos de imobilizado únicos: ${totalCodigos}`);
        console.log(`👉 Total de itens (sub-ativos) processados: ${totalItens}`);
        console.log(`👉 Total de linhas ignoradas (dados inválidos): ${linhasIgnoradas}`);
        
        if (codigosInvalidos.size > 0) {
            const amostra = Array.from(codigosInvalidos).slice(0, 5).join(', ');
            console.log(`⚠️  Amostra de códigos inválidos ignorados: ${amostra}...`);
        }
        
        console.log("-".repeat(40));
        console.log(`\n🚀 Arquivo JSON gerado com sucesso em:`);
        console.log(caminhoArquivoSaida);
        console.log("=".repeat(40));
        
    } catch (error) {
        console.log("\n❌ Ocorreu um erro inesperado durante o processamento.");
        console.log(`   Tipo do erro: ${error.constructor.name}`);
        console.log(`   Detalhes: ${error.message}`);
    } finally {
        await aguardarEntrada("\nPressione Enter para sair...");
    }
}

// Executa o script se for chamado diretamente
if (require.main === module) {
    processarDados();
}

module.exports = { processarDados, formatarValor, formatarData };

